src/sleep/sleep.c

```diff
@@ -38,8 +39,192 @@
+                        goto fail_cleanup;
+                }
+
+                /* Nothing to process */
+                if (fiemap->fm_mapped_extents == 0)
+                        break;
+
+                /* Result fiemap have to hold all the extents for the hole file */
+
+                /* Read in the extents */
+                extents_size = sizeof(struct fiemap_extent) *
+                                      (fiemap->fm_mapped_extents);
+
+                /* Resize fiemap to allow us to read in the extents */
+                fm_tmp = realloc(fiemap,
+                                 sizeof(struct fiemap) + extents_size);
+                if (!fm_tmp) {
+                        fprintf(stderr, "Out of memory reallocating fiemap\n");
+                        goto fail_cleanup;
+                }
+                fiemap = fm_tmp;
+
+                memset(fiemap->fm_extents, 0, extents_size);
+                fiemap->fm_extent_count = fiemap->fm_mapped_extents;
+                fiemap->fm_mapped_extents = 0;
+
+                if (ioctl(fd, FS_IOC_FIEMAP, fiemap) < 0) {
+                        fprintf(stderr, "fiemap ioctl() FS_IOC_FIEMAP failed, errno=%d (%s)\n",
+                                errno, strerror(errno));
+                        goto fail_cleanup;
+                }
+
+                extents_size = sizeof(struct fiemap_extent) *
+                                      (result_extents +
+                                       fiemap->fm_mapped_extents);
+
+                /* Resize result_fiemap to allow us to read in the extents */
+                fm_tmp = realloc(result_fiemap,
+                                 sizeof(struct fiemap) + extents_size);
+                if (!fm_tmp) {
+                        fprintf(stderr, "Out of memory allocating fiemap\n");
+                        goto fail_cleanup;
+                }
+                result_fiemap = fm_tmp;
+
+                memcpy(result_fiemap->fm_extents + result_extents,
+                       fiemap->fm_extents,
+                       sizeof(struct fiemap_extent) *
+                       fiemap->fm_mapped_extents);
+
+                result_extents += fiemap->fm_mapped_extents;
+
+                /* Highly unlikely that it is zero */
+                if (fiemap->fm_mapped_extents) {
+                        const __u32 i = fiemap->fm_mapped_extents - 1;
+
+                        fiemap_start = fiemap->fm_extents[i].fe_logical +
+                                       fiemap->fm_extents[i].fe_length;
+
+                        if (fiemap->fm_extents[i].fe_flags & FIEMAP_EXTENT_LAST)
+                                break;
+                }
+        }
+
+        result_fiemap->fm_mapped_extents = result_extents;
+        free(fiemap);
+
+        return result_fiemap;
+
+fail_cleanup:
+        if (result_fiemap)
+                free(result_fiemap);
+
+        if (fiemap)
+                free(fiemap);
```

Comment:
usually, we rely on `_cleanup_free_` for making such clean-up paths automatic. please use that here too (moreover free() is actually fine with NULL parameters, hence you could drop the explicit checks anyway)